In [15]:
import pandas as pd

In [20]:
! head -500 ~/shared-data/clickhouse_data/friends.json > ./friends_sample.json

! head -500 ~/shared-data/clickhouse_data/followers.json > ./followers_sample.json

! head -500 ~/shared-data/clickhouse_data/likes.json > ./likes_sample.json

! head -500 ~/shared-data/clickhouse_data/posts.json > ./posts_sample.json

! head -500 ~/shared-data/clickhouse_data/user_profiles.json > ./user_profiles_sample.json

## Data exploration

In [23]:
! head -5 ./friends_sample.json

{"user_id":"6","friend_id":"197","ctime":"2020-03-03 18:25:31"}
{"user_id":"6","friend_id":"2675","ctime":"2020-03-03 18:25:31"}
{"user_id":"6","friend_id":"2807","ctime":"2020-03-03 18:25:31"}
{"user_id":"6","friend_id":"5333","ctime":"2020-03-03 18:25:31"}
{"user_id":"6","friend_id":"7373","ctime":"2020-03-03 18:25:31"}


In [21]:
friends_df = pd.read_json("./friends_sample.json", lines=True)
friends_df.head(5)

user_id  friend_id                ctime
0        6        197  2020-03-03 18:25:31
1        6       2675  2020-03-03 18:25:31
2        6       2807  2020-03-03 18:25:31
3        6       5333  2020-03-03 18:25:31
4        6       7373  2020-03-03 18:25:31

In [24]:
! head -5 ./followers_sample.json

{"profile_id":"328582","follower_id":"4652928","ctime":"2020-03-07 10:09:50"}
{"profile_id":"328582","follower_id":"5638400","ctime":"2020-03-07 10:09:50"}
{"profile_id":"328582","follower_id":"5800954","ctime":"2020-03-07 10:09:50"}
{"profile_id":"328582","follower_id":"6275015","ctime":"2020-03-07 10:09:50"}
{"profile_id":"328582","follower_id":"6530002","ctime":"2020-03-07 10:09:50"}


In [22]:
followers_df = pd.read_json("./followers_sample.json", lines=True)
followers_df.head(5)

profile_id  follower_id                ctime
0      328582      4652928  2020-03-07 10:09:50
1      328582      5638400  2020-03-07 10:09:50
2      328582      5800954  2020-03-07 10:09:50
3      328582      6275015  2020-03-07 10:09:50
4      328582      6530002  2020-03-07 10:09:50

In [25]:
! head -5 ./likes_sample.json

{"item_type":"post","owner_id":"-174416405","item_id":"1083","liker_id":"72689","ctime":"0000-00-00 00:00:00","like_date":"0000-00-00 00:00:00","post_date":"2019-03-15 07:34:00"}
{"item_type":"post","owner_id":"-174416405","item_id":"1103","liker_id":"72689","ctime":"0000-00-00 00:00:00","like_date":"0000-00-00 00:00:00","post_date":"2019-03-17 08:05:00"}
{"item_type":"post","owner_id":"-170321944","item_id":"328776","liker_id":"72689","ctime":"0000-00-00 00:00:00","like_date":"0000-00-00 00:00:00","post_date":"2019-03-11 23:50:00"}
{"item_type":"post","owner_id":"-170321944","item_id":"334079","liker_id":"72689","ctime":"0000-00-00 00:00:00","like_date":"0000-00-00 00:00:00","post_date":"2019-03-13 22:15:00"}
{"item_type":"post","owner_id":"-168996002","item_id":"19811","liker_id":"72689","ctime":"0000-00-00 00:00:00","like_date":"0000-00-00 00:00:00","post_date":"2019-03-05 19:43:00"}


In [26]:
likes_df = pd.read_json("./likes_sample.json", lines=True)
likes_df.head(5)

item_type   owner_id  item_id  liker_id                ctime  \
0      post -174416405     1083     72689  0000-00-00 00:00:00   
1      post -174416405     1103     72689  0000-00-00 00:00:00   
2      post -170321944   328776     72689  0000-00-00 00:00:00   
3      post -170321944   334079     72689  0000-00-00 00:00:00   
4      post -168996002    19811     72689  0000-00-00 00:00:00   

             like_date            post_date  
0  0000-00-00 00:00:00  2019-03-15 07:34:00  
1  0000-00-00 00:00:00  2019-03-17 08:05:00  
2  0000-00-00 00:00:00  2019-03-11 23:50:00  
3  0000-00-00 00:00:00  2019-03-13 22:15:00  
4  0000-00-00 00:00:00  2019-03-05 19:43:00

In [27]:
! head -5 ./posts_sample.json

{"ctime":"0000-00-00 00:00:00","date":"2012-03-13 00:00:00","post_id":"5","from_id":"78062","owner_id":"-36559458","comments_count":-1,"likes_count":2,"reposts_count":0,"views_count":0,"text":"","signed_by":"-1","post_type":"post","reposted_from_owner_id":null,"reposted_from_post_id":null,"geo":0,"geo_lat":null,"geo_lon":null,"geo_hash":null,"photo_attachments.owner_id":[],"photo_attachments.photo_id":[],"photo_attachments.size":[],"photo_attachments.url":[],"photo_attachments.geo_lat":[],"photo_attachments.geo_lon":[],"photo_attachments.geo_hash":[],"video_attachments.owner_id":[],"video_attachments.video_id":[],"video_attachments.views":[],"audio_attachments.owner_id":[],"audio_attachments.audio_id":[],"audio_attachments.artist":[],"audio_attachments.title":[],"doc_attachments.owner_id":[],"doc_attachments.doc_id":[],"doc_attachments.title":[],"doc_attachments.size":[],"doc_attachments.url":[],"doc_attachments.type":[],"link_attachments.title":[],"link_attachments.url":[],"page_attac

In [29]:
posts_df = pd.read_json("./posts_sample.json", lines=True)
posts_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 50 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   ctime                           500 non-null    object        
 1   date                            500 non-null    datetime64[ns]
 2   post_id                         500 non-null    int64         
 3   from_id                         500 non-null    int64         
 4   owner_id                        500 non-null    int64         
 5   comments_count                  500 non-null    int64         
 6   likes_count                     500 non-null    int64         
 7   reposts_count                   500 non-null    int64         
 8   views_count                     500 non-null    int64         
 9   text                            500 non-null    object        
 10  signed_by                       500 non-null    int64         
 11  post_t

In [36]:
user_profiles_df = pd.read_json("./user_profiles_sample.json", lines=True)
user_profiles_df.head(5)

ctime        id first_name last_name          screen_name  \
0  2019-01-14 14:43:44  15588946     Karina   Enokyan           id15588946   
1  2019-01-14 14:43:45  15590249      Sanya     Krupy                 None   
2  2019-01-14 14:43:45  15590390    Natalya   Ershova            tymannaya   
3  2019-01-14 14:43:45  15591560      Alina    Orlova  loveyoumybestfriend   
4  2019-01-14 14:43:45  15592275      Elena     Gutor           id15592275   

  maiden_name nickname      bdate  birth_date  sex  ...  schools.year_to  \
0                            20.1         NaN    1  ...               []   
1        None     None       None         NaN    2  ...               []   
2                           23.10         NaN    1  ...               []   
3                       10.8.1995  19950810.0    1  ...               []   
4     Kanaeva           17.5.1984  19840517.0    1  ...               []   

   universities.chair  universities.chair_name  universities.education_form  \
0                  []                       []                           []   
1                  []                       []                           []   
2              [None]                   [None]          [Distance Learning]   
3                  []                       []                           []   
4                  []                       []                           []   

  universities.education_status  universities.faculty  \
0                            []                    []   
1                            []                    []   
2                        [None]                [1441]   
3                            []                    []   
4                            []                    []   

                       universities.faculty_name  universities.graduation  \
0                                             []                       []   
1                                             []                       []   
2  [Факультет журналистики (Институт массмедиа)]                   [None]   
3                                             []                       []   
4                                             []                       []   

  universities.id universities.name  
0              []                []  
1              []                []  
2           [328]            [РГГУ]  
3              []                []  
4              []                []  

[5 rows x 102 columns]

## Report

### MVs:
* Total number of likes by user/group
* Number of comments in user's/group's posts by months

### Total number of likes by user/group

##### Main local table

Here we sort our index by liker_id to efficiently retrieve groups of likes per user.  

    CREATE TABLE IF NOT EXISTS likes_loc ON CLUSTER kube_clickhouse_cluster
    (
        item_type String,
        owner_id Int64,
        item_id Int64,
        liker_id Int64,
        ctime DateTime,
        like_date DateTime,
        post_date DateTime
    )
    ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/likes', '{replica}')
    PARTITION BY toYear(post_date)
    ORDER BY (liker_id);

##### Local MV

    CREATE MATERIALIZED VIEW IF NOT EXISTS likes_per_user_loc ON CLUSTER kube_clickhouse_cluster
    ENGINE = SummingMergeTree()
    ORDER BY (liker_id)
    POPULATE
    AS SELECT
      liker_id,
      count(item_id) AS number_of_likes
    FROM likes_loc
    GROUP BY liker_id;

##### Wrap local tables into Distributed. Specify hash for the main table as we will insert via Distributed.

    CREATE TABLE IF NOT EXISTS likes ON CLUSTER kube_clickhouse_cluster AS likes_loc
    ENGINE = Distributed(kube_clickhouse_cluster, default, likes_loc, xxHash64(liker_id));

    CREATE TABLE IF NOT EXISTS likes_per_user ON CLUSTER kube_clickhouse_cluster AS likes_per_user_loc
    ENGINE = Distributed(kube_clickhouse_cluster, default, likes_per_user_loc);

##### Kafka consumers

    CREATE TABLE IF NOT EXISTS likes_queue ON CLUSTER kube_clickhouse_cluster
    (
        item_type String,
        owner_id Int64,
        item_id Int64,
        liker_id Int64,
        ctime DateTime,
        like_date DateTime,
        post_date DateTime
    )
    ENGINE = Kafka('kafka-svc:9092', 'likes-topic', 'group1', 'JSONEachRow');

    CREATE MATERIALIZED VIEW IF NOT EXISTS likes_consumer TO likes AS SELECT * FROM likes_queue;

##### Final query

    SELECT liker_id, sum(number_of_likes) FROM likes_per_user GROUP BY liker_id;

##### Add some data

    cd $KAFKA_DIR
    head -n 1000000 /shared-data/clickhouse_data/likes.json | ./kafka-console-producer.sh --bootstrap-server localhost:9092 --topic likes-topic &> /dev/null

### Number of comments in user's/group's posts by months

##### Main local table

We firstly sort index by date and then by from_id.

    CREATE TABLE IF NOT EXISTS posts_loc ON CLUSTER kube_clickhouse_cluster
    (
        post_id Int64,
        from_id Int64,
        comments_count Int64,
        date DateTime
    )
    ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/posts', '{replica}')
    ORDER BY (date, from_id);

##### Local MV

    CREATE MATERIALIZED VIEW IF NOT EXISTS comments_per_user_loc ON CLUSTER kube_clickhouse_cluster
    ENGINE = SummingMergeTree()
    ORDER BY (year, month, user_id)
    POPULATE
    AS SELECT
      from_id AS user_id,
      toYear(date) AS year,
      toMonth(date) AS month,
      sum(comments_count) AS total_comments
    FROM posts_loc
    WHERE comments_count > 0
    GROUP BY year, month, from_id;

##### Wrap local tables into Distributed. Specify hash for the main table as we will insert via Distributed.

    CREATE TABLE IF NOT EXISTS posts ON CLUSTER kube_clickhouse_cluster AS posts_loc
    ENGINE = Distributed(kube_clickhouse_cluster, default, posts_loc, xxHash64(toYear(date)));

    CREATE TABLE IF NOT EXISTS comments_per_user ON CLUSTER kube_clickhouse_cluster AS comments_per_user_loc
    ENGINE = Distributed(kube_clickhouse_cluster, default, comments_per_user_loc);
    
##### Kafka consumers

    CREATE TABLE IF NOT EXISTS posts_queue ON CLUSTER kube_clickhouse_cluster
    (
        post_id Int64,
        from_id Int64,
        comments_count Int64,
        date DateTime
    )
    ENGINE = Kafka('kafka-svc:9092', 'posts-topic', 'group-1', 'JSONEachRow');

    CREATE MATERIALIZED VIEW IF NOT EXISTS posts_consumer ON CLUSTER kube_clickhouse_cluster TO posts AS SELECT * FROM posts_queue;

##### Final query

    SELECT year, month, user_id, sum(total_comments)
    FROM comments_per_user
    GROUP BY year, month, user_id;

##### Add some data

    cd $KAFKA_DIR
    head -n 1000000 /shared-data/clickhouse_data/posts.json | ./kafka-console-producer.sh --bootstrap-server localhost:9092 --topic posts-topic &> /dev/null

### Table definitions

##### Friends

"user_id":"6","friend_id":"197","ctime":"2020-03-03 18:25:31"

    CREATE TABLE IF NOT EXISTS friends_loc ON CLUSTER kube_clickhouse_cluster
    (
        user_id Int64,
        friend_id Int64,
        ctime DateTime
    )
    ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/friendss', '{replica}')
    ORDER BY (user_id, friend_id);

    CREATE TABLE IF NOT EXISTS friends ON CLUSTER kube_clickhouse_cluster AS friends_loc
    ENGINE = Distributed(kube_clickhouse_cluster, default, friends_loc, user_id);

    CREATE TABLE IF NOT EXISTS friends_queue ON CLUSTER kube_clickhouse_cluster AS friends_loc
    ENGINE = Kafka('kafka-svc:9092', 'friends-topic', 'group1', 'JSONEachRow');

    CREATE MATERIALIZED VIEW IF NOT EXISTS friends_consumer ON CLUSTER kube_clickhouse_cluster TO friends AS SELECT * FROM friends_queue;

##### Followers

"profile_id":"328582","follower_id":"4652928","ctime":"2020-03-07 10:09:50"

    CREATE TABLE IF NOT EXISTS followers_loc ON CLUSTER kube_clickhouse_cluster
    (
        profile_id Int64,
        follower_id Int64,
        ctime DateTime
    )
    ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/followers', '{replica}')
    ORDER BY (profile_id, follower_id);

    CREATE TABLE IF NOT EXISTS followers ON CLUSTER kube_clickhouse_cluster AS followers_loc
    ENGINE = Distributed(kube_clickhouse_cluster, default, followers_loc, profile_id);

    CREATE TABLE IF NOT EXISTS followers_queue ON CLUSTER kube_clickhouse_cluster AS followers_loc
    ENGINE = Kafka('kafka-svc:9092', 'followers-topic', 'group1', 'JSONEachRow');

    CREATE MATERIALIZED VIEW IF NOT EXISTS followers_consumer ON CLUSTER kube_clickhouse_cluster TO followers AS SELECT * FROM followers_queue;

##### User Profiles

    CREATE TABLE IF NOT EXISTS user_profiles_loc ON CLUSTER kube_clickhouse_cluster
    (
        id Int64,
        first_name String,
        last_name String,
        nickname Nullable(String),
        sex Nullable(Int8),
        deactivated Nullable(Int8),
        verified Nullable(Int8),
        followers_count Nullable(Int64),
        status Nullable(String),
        city_id Nullable(Int64),
        city_title Nullable(String),
        country_id Nullable(Int64),
        country_title Nullable(String),
        ctime DateTime
    )
    ENGINE = ReplicatedMergeTree('/clickhouse/tables/{shard}/user_profiles', '{replica}')
    ORDER BY (id);

    CREATE TABLE IF NOT EXISTS user_profiles ON CLUSTER kube_clickhouse_cluster AS user_profiles_loc
    ENGINE = Distributed(kube_clickhouse_cluster, default, user_profiles_loc, id);

    CREATE TABLE IF NOT EXISTS user_profiles_queue ON CLUSTER kube_clickhouse_cluster AS user_profiles_loc
    ENGINE = Kafka('kafka-svc:9092', 'user-profiles-topic', 'group1', 'JSONEachRow');

    CREATE MATERIALIZED VIEW IF NOT EXISTS user_profiles_consumer ON CLUSTER kube_clickhouse_cluster TO user_profiles AS SELECT * FROM user_profiles_queue;
